### PR TITLE
Fix qual pushdown through Set Operation with window functions.

### LINKED ISSUE #5762: Predicate incorrectly pushed down through UNION ALL with window functions
### Greenplum version or build

Master, 5X_STABLE

### Expected behavior

```
select * from (
  select i, row_number() over (partition by i/3) from generate_series(1, 5) i
  UNION ALL
  select i, row_number() over (partition by i/3) from generate_series(1, 5) i
) as sq WHERE i = 3;
 i | row_number 
---+------------
 3 |          3
 3 |          3
(2 rows)
```

### Actual behavior

```
 i | row_number 
---+------------
 3 |          1
 3 |          1
(2 rows)
```

The planner incorrectly pushes down the `i = 3` qual through the UNION ALL, and below the window function evaluation. So the row_number() window function only sees the row with `i = 3`, which is wrong. The qual is supposed to be evaluated after the window function.

### PR DESCRIPTION
**Pull request by guofengrichard**
Previously when checking if it is safe to push down a qual through
window functions in qual_is_pushdown_safe(), we did not consider the
case with Set Operations. As a result, qual would be incorrectly pushed
down through Set Operations with window functions.

In this patch, we recurse through setOperations tree when check for
window functions.

This patch fixes #5762 .

Co-authored-by: Melanie Plageman <mplageman@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`

### CODE REVIEW

**Comment by guofengrichard** [comment]
> Considering that we only traversed one layer down in the setOperations parameter, using recursive functions is a bit confusing. I suggest using a simple qual_is_pushdown_safe_window_func function that checks the query's windowClause only in the function. This function is called in qual_is_pushdown_safe and recurse_qual_pushdown_safe_window_func. This way we don't need recursive calls, nor do we need to pass topquery as a stop condition.

Hmm, actually we need to recursively traverse the setOperations several layers down, like in the case `sq1 UNION ALL sq2 UNION ALL sq3 ...`.

This behaves in the same way as the function `subquery_is_pushdown_safe()` and  `recurse_pushdown_safe()` from upstream.

**Comment by khannaekta** [comment]
Can we add the sql from the github issue, as a test.

File: src/backend/optimizer/path/allpaths.c [hunk 1]
```diff
@@ -1757,6 +1660,95 @@ qual_is_pushdown_safe(Query *subquery, RangeTblEntry *rte, Index rti, Node *qual
...
+			return false;
+		if (!recurse_qual_pushdown_safe_window_func(op->rarg, topquery, resno))
+			return false;
+	}
+	else
+	{
+		elog(ERROR, "unrecognized node type: %d",
+			 (int) nodeTag(setOp));
+	}
+	return true;
```
Comment (khannaekta) [inline src/backend/optimizer/path/allpaths.c]: In what scenario will we `return true` from here? I am trying to come up with a query but couldn't think of any.

Comment (guofengrichard) [reply src/backend/optimizer/path/allpaths.c]: I don't think any code path will reach this line. It is here to keep compiler quiet.

**Comment by guofengrichard** [comment]
> Can we add the sql from the github issue, as a test.

There is a same test query in test case `olap_window_seq`:
```
explain with CTE as (select i, row_number() over (partition by j) j from window_preds union all select i, row_number() over (partition by j) from window_preds) select * from cte where i = 1;
```

Previously the output in expect file for this query is wrong, that's why we did not catch this issue for so long time.

File: src/backend/optimizer/path/allpaths.c (hunk 1 above)
Comment (melanieplageman) [reply src/backend/optimizer/path/allpaths.c]: What version is your compiler?

Comment (guofengrichard) [reply src/backend/optimizer/path/allpaths.c]: I am using gcc 6.2.0, but IMO a warning type of `-Wreturn-type` will always appear without this `return true`, no matter what compiler version you are using.

**Pull request merged**